Add warning to never use Random.get_state as input to v4_gen and make the example safe on OCaml 5

edwintorok opened the pull request:
Due to the deprecation of `Uuidm.v` I've seen a PR on https://github.com/xapi-project/xen-api that used random state incorrectly, in a way that would've resulted in identical Uuids in the same program.

Although this was immediately obvious to me when reviewing the PR (having dealt with the Random module in detail in the past), lets improve the documentation so that it is also more obvious for everyone.
I've also opened a PR on https://github.com/ocaml/ocaml/pull/13533 to make the standard library's documentation more explicit.

In fact it wasn't until OCaml 5 that it was documented that `Random.get_state` returns a copy, previously it (incorrectly) said that it returns the current state, which wasn't true (if it returned the current state, then it would've been updated by each request to generate random bits, and would've therefore generated unique values, at least in a single threaded program).

edwintorok committed:
Add warning to never use Random.get_state as input to v4_gen

Unless you really want identical UUIDs...

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
In fact on OCaml 5 one should be very careful to avoid accesses from multiple domains to the same state (i.e. make a fresh call to `Random.split ()`, instead of using partial application).

The example given in `index.mld` is not safe for OCaml 5 (even though OCaml5's Random module is Domain safe, it is only domain-safe when you use its own functions on the internal state, which uses DLS to ensure each domain has its own, or if you always obtain a fresh split or freshly created one).

AFAICT OCaml 4.14's Random.bits was thread safe because it avoided calling anything that might've given OCaml an opportunity to switch threads (although the documentation didn't give such guarantees, consulting the generated assembly seems to confirm this).
However on OCaml 5 accessing the same State.t from multiple domains is not safe: there are at least 3 separate accesses to the state, not protected by any atomic or mutex operations. Accessing it from multiple threads within the same domain is safe for the same reason it was safe on OCaml 4: no possibility for OCaml to switch threads there as the runtime lock is not released.

I think a simple fix would be to update the example to use `Random.split ()` without partial application, which would ensure we get a State.t that is not shared with any other domain (calling `make_self_init ()` would be too expensive).
Or one could use mutexes (and that'll work on both OCaml 4 and OCaml 5), but we need to find some solution for Uuidm's documentation that is  multi-domain safe.

edwintorok committed:
doc(index.mld): Avoid generating duplicate UUIDs on OCaml 5

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

dbuenzli commented:
It would have been nice to have a version that's independent from the ocaml version but I don't think that's possible (except by using mutexes but then you have to compile with the threads library on 4).

It's a bit silly `Random.get_state` does not do what it seems to do (by reading I would have thought it gives you the ability to use the default implicit random state with functions that asks for a random state). If it was not copying we could provide this code which would work across versions:

```ocaml
let () = Random.self_init () 
let uuidm () =  Uuidm.v4_gen (Random.get_state ()) ()
```
But we can't :–)

However I'm a bit concerned at how you use `split` here, is it really made to be called so often ? I draws 4 int64 and allocates a new state just for a few random calls made by `v4_gen`. Should we not rather show how to use domain local storage ? E.g. ? 

```ocaml
let rstate = Domain.DLS.new_key ~split_from_parent:State.split Random.State.self_init
let uuidm () = Uuidm.v4_gen (Domain.DLS.get rstate) ()
```

Perhaps @gasche can comment ?

edwintorok commented:
I haven't measured whether `Random.split` or DLS is faster, but I'd expect that you are correct and DLS is.
Although as long as we can avoid reading from `/dev/urandom` for my purposes that is "fast enough", I'm more concerned about correctness (i.e. UUIDs actually have to be unique as their name says).

One solution that would be compatible with both OCaml 4 and OCaml 5 would be to use this OPAM package: https://ocaml.org/p/stdlib-random/latest/doc/Random5/index.html, which makes `Random5.split` available on OCaml 4 too.
Although DLS can also be used on OCaml 4 with domain-shims.

A mutex would also work (and in fact is the solution I went for in the short term, but I'll likely switch to one of the other solutions soon):
```
let make_uuid_fast =
  let uuid_state = Random.State.make_self_init () in
  (* On OCaml 5 we could use Random.State.split instead,
     and on OCaml 4 the mutex may not be strictly needed,
    but this allows the same code to be compatible with both
  *)
  let m = Mutex.create () in
  let finally () = Mutex.unlock m in
  let gen = Uuidm.v4_gen uuid_state in
  fun () -> Mutex.lock m ; Fun.protect ~finally gen
```

(this takes ~90ns on my test machine vs ~1370ns if I read from /dev/urandom every time, I'll try to measure the other proposed solutions later)

gasche commented:
My impression is that it is a usability problem that there is no way to get the current `Random.State.t` without making a copy of it. Having that feature available would let library authors take a Random.State.t parameter, and also program authors use the global per-domain state instead of handling state themselves. This is something that could be reported upstream, I think.

I see two mitigation strategies:
1. Libraries such as `uuid` could have a variant that depends on a `(unit -> int64)` function instead of a Random.State.t parameter. In fact, knowing @dbuenzli's taste, I am almost surprised that this is not already the case. Users could then pass `(fun () -> Random.State.bits64 my_state)`, or `Random.bits64`.
2. Program authors could implement their own global PRNG state using Random.State + DLS. This is in fact rather easy.

Remarks:
1. Calling `Random.split ()` concurrently is a bit weird as calls from several threads may be interleaved on the same random state. I don't think this breaks the implementation of `split ()` or weakens its statistical properties, it is just a bit weird.
2. I agree that using a mutex is reasonable (5.1 and later have `Mutex.protect` to makes this easier)

dbuenzli commented:
Thanks @gasche for your input.

> 1. Libraries such as `uuid` could have a variant that depends on a `(unit -> int64)` function instead of a Random.State.t parameter.

Note somehow it has. You have [base constructors](https://erratique.ch/software/uuidm/doc/Uuidm/index.html#uuids) with full control to implement your own generators (e.g. to use a CSRPNG). I see the `Random` based [generators](https://erratique.ch/software/uuidm/doc/Uuidm/index.html#gen) as a convenience.

> 2\. Program authors could implement their own global PRNG state using Random.State + DLS. This is in fact rather easy.

I think this is the second example [here](https://github.com/dbuenzli/uuidm/pull/15#issuecomment-2400009056). 

@edwintorok I don't think it's worth bothering people with 4 and 5. At the rate where this package gets releases I'd rather just have the cleanest example for 5. either with `Mutex.protect` or DLS, though the status of the latter seems [still hanging](https://github.com/ocaml/ocaml/discussions/13356).

edwintorok commented:
It seems to me that we're overcomplicating this a bit (having to use DLS, or external Random modules, or requiring a new version of runtime that has `Random.state ()`), where the problem seems to have been all along that `v4_gen` takes a `Random.State.t` as input. That may have been a fine choice for OCaml 4 (especially given that it didn't document that it was giving you a copy)

I thought and discarded this convenience function:
```
+let v4_gen' () =
+  let r0 = Random.bits64 () in
+  let r1 = Random.bits64 () in
+  let u = Bytes.create 16 in
+  Bytes.set_int64_be u 0 r0;
+  Bytes.set_int64_be u 8 r1;
+  make u ~version:4
+
```

This does not give the user a choice on how to initialize State.t (other than initializing the global one), but it is a lot harder to misuse: 
* you can call it from multiple threads safely
* you can call it from multiple domains safely (`Random.bits64` internally uses DLS)
* you can trivially "partially apply" it (well it is just an alias for `v4_gen'` at that point) 

Although there is still one mistake you can make:
* if you forget to initialize the global Random module, you'll get back the same random numbers on each program run. 

Unfortunately there is no way to tell whether Random was initialized or not, and a library reinitializing Random wouldn't be a good idea either (maybe the user took great care in providing high quality entropy).
And that error would be worse than the current one, where you only get a duplicate UUID on a very unlikely race condition: in this case you'd always get duplicate UUIDs on each program run.

But this can be improved with some conditional compilation: on OCaml 5 use the DLS snippet you've shown, and on OCaml 4 keep using `make_self_init`:

OCaml 4:
```
let rstate = Random.State.self_init ()
let v4_gen' () = v4_gen rstate ()
```

OCaml 5:
```
let rstate = Domain.DLS.new_key ~split_from_parent:State.split Random.State.self_init
let v4_gen' () = v4_gen (Domain.DLS.get rstate) ()
```

(`v4_gen'` is just an example, we'd need a better name).

I think it'd be important to have a convenience function that always works correctly as part of the library: if the library can't supply one, then how could a user write a correct one?

(BTW for several years XAPI has been using Uuidm in a lot more paranoid way: always read some bytes from `/dev/urandom` and use that to construct UUIDs.
That is very slow, but is very very unlikely to generate duplicate UUIDs (unless your OS is buggy, e.g. you're generating it inside a VM that you've forked -- e.g. restored to an earlier in-memory snapshot --, and the VM's OS didn't implement the necessary generation ID detection logic to reseed its RNG.
It is only more recently that I've looked at using OCaml's RNG to seed the generation of UUIDs.
).

(also having a convenience function as proposed in point 1 https://github.com/dbuenzli/uuidm/pull/15#issuecomment-2400683637 would be harder to misuse, but if we're providing convenience functions, why not provide one that has no known faults?)

dbuenzli commented:
> where the problem seems to have been all along that `v4_gen` takes a `Random.State.t` as input. That may have been a fine choice for OCaml 4 (especially given that it didn't document that it was giving you a copy)

The fact that you could take a `Random.State.t` was not so that you could wrongly use `Random.get_state`. 

It was so that the library would not side effect on the default state (which I would consider a sin) and allow to control this yourself with your own state (including possibly using this for testing, though that is a bit hindered by the fact that upstream likes to change the algorithm behind your back). 

In general except for short tests I never use the default state (and I don't think libraries should either) as I worry a lot about controlling mutable state and effects in my programs in explicit manners.

> I think it'd be important to have a convenience function that always works correctly as part of the library: if the library can't supply one, then how could a user write a correct one?

Well I would actually rather find it important for `Random.State.t` to be composable in OCaml 5 without needing *for each library* and user to perform heaving duty gymnastics or (for the reasons mentioned previously) relying on the default state. 

Perhaps it would be a good place for the `Stdlib` to start providing multicore safe abstractions. I note that it was not marked as unsynchronized unsafe and even worse it has this bit in the docs:

> This allows using one or several deterministic PRNGs, even in a multi-threaded program, without interference from other parts of the program.

So if a mutex is needed I think it should be added by `Stdlib`. A lot of care is put into making `Random` a good generator it would be silly if all that was thwarted for stupid usability reasons.
